### DISCUSSION TITLE
How to save objects from fabric.js without anti-aliasing?

### DISCUSSION THREAD

**Discussion by alanwilter** [fabricjs/fabric.js#9722] [Q&A]
First, I'm not a JS developer, I'm just trying to fix an issue that we inherited from the JS developer who left our team.

I have this example to help:
[view.html][1]
[fabric.json][2]

The problem is that when I save the canvas in PNG, both the square and the other objects are saved with anti-aliasing.

I'm looking for a way to save these objects with sharp colour contrast, no anti-aliasing at all.

[![enter image description here][3]][3]

  [1]: https://gist.github.com/alanwilter/178f5edcf98b18cdc0498dd5b305724e
  [2]: https://gist.github.com/alanwilter/365335311add365f8e4f6c1b09b8dae7
  [3]: https://i.stack.imgur.com/Rsqpz.png

**Comment by asturur** [comment]
Hi @alanwilter from the screenshot you post is difficult to understand if the antialiasing you don't want is just blurriness or proper anti aliasing.

Since your profile description i m going to assume you know what anti-aliasing is and sadly there is no way to disable it in canvas. It is embedded in the drawing routines of the canvas and we can't do much about it.

It should be possible to output a perfect aligned rect without any anti aliasing, but not for the irregular shapes that have curves or for diagonal lines.

If you have few objects and you want a small canvas you may consider using point in polygon function we have, check for every point of the canvas if it is inside or outside the shape and draw it with a solid color.
It will be painfully slow but is better than nothing.

**Comment by ShaMan123** [comment]
From what I can tell it is due to object caching context isolation caveat
You are using the eraser, it needs context isolation, therefore object caching is turned on. 
That produces bad visuals due to wrong scale accuracy resulting in blur because there is not enough pixel data
#9693

**Reply by ShaMan123** [reply]
~~Provide a live demo and try to work on the PR branch to see if that solves your issue~~

Strike that, eraser is not part of v6

**Reply by ShaMan123** [reply]
This is the PR's juice
https://github.com/fabricjs/fabric.js/blob/4ef210c62e4a8a90e0868490785efd15d904481b/src/shapes/Object/Object.ts#L796-L829

**Reply by ShaMan123** [reply]
you will need to use https://github.com/fabricjs/fabric.js/blob/4ef210c62e4a8a90e0868490785efd15d904481b/src/canvas/CanvasProvider.ts#L49

**Reply by asturur** [reply]
So is not anti-aliasing what he is talking about?

**Reply by alanwilter** [reply]
The example I posted does have eraser indeed, but the example html visualisation won't render that.
I have thousands of images like those that we use in our annotation web application.
The images are then converted in B&W for ML trainings and when taking from fabric object to PNG the images are coming with anti-aliasing, which is not the end of the world but affects e.g. Dice score calculations depending on how we apply a threshold to the grayscale levels of the anti-aliasing.

**Reply by asturur** [reply]
Can you define anti-aliasing please?

For me this is it:

![image](https://github.com/fabricjs/fabric.js/assets/1194048/76d8a536-1e3f-476a-a8e4-38d9e6fcedaa)

and this can't be disabled.

**Reply by alanwilter** [reply]
Precisely, I'm looking for the without anti-aliasing images. 

So if this can't be achieved, patience.

**Reply by asturur** [reply]
It can't be achieved with canvas configuration because it has that strictly embedded.
I think it can be tweaked under node if you are rendering in nodeJS, and it can be done manually for polygons and geometric shapes if you want to write some code, but definetely not for text in a browser or for paths.

**Reply by ShaMan123** [reply]
Try imageSmoothing
